{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tidy data in Pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handy hints "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some Jupyter shortcut keys:\n",
    "\n",
    "- Esc : get into Command mode (leaves Edit mode)\n",
    "- Enter : edit a cell (puts you in Edit mode)\n",
    "- h : see help (see all commands)\n",
    "- Shift+Enter or Ctrl+Enter : run the code in the cell\n",
    "- a / b : add new cell above/below\n",
    "- m : turn current cell into a Markdown cell\n",
    "- y : turn current cell into a Code cell\n",
    "- 1-6 : turn current cell into a heading (a type of Markdown cell)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can get help on Python objects and functions with `help()` or the `?` operator."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-19T11:29:49.600774Z",
     "start_time": "2019-01-19T11:29:49.581459Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-19T11:29:50.000633Z",
     "start_time": "2019-01-19T11:29:49.987260Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# This causes Jupyter to display any matplotlib plots directly in the notebook\n",
    "# It also works for pandas and seaborn, since they use matplotlib to render plots\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-19T11:29:51.169379Z",
     "start_time": "2019-01-19T11:29:51.163406Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# pyplot (as plt) is the module we'll primarily use to instantiate matplotlib plot objects\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pandas "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can find the Pandas website and documentation at https://pandas.pydata.org/\n",
    "\n",
    "Pandas is a popular Python library for handling tabular data. It provides much of the same functionality for Python as do data frames in the R language. \n",
    "\n",
    "The fundamental data types in Pandas are a Series, representing a 1D array of data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      Diego\n",
       "1    Jessica\n",
       "2      Farah\n",
       "dtype: object"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = pd.Series([\"Diego\",\"Jessica\",\"Farah\"])\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and a DataFrame, representing a 2D table of data: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Diego</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Jessica</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Farah</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Name  Age\n",
       "0    Diego   34\n",
       "1  Jessica   27\n",
       "2    Farah   50"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame({'Name': [\"Diego\",\"Jessica\",\"Farah\"],\n",
    "                   'Age': [34, 27, 50]})\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can think of a DataFrame like a spreadsheet or a table in a database; every column represents a variable in the dataset. Each column of a DataFrame has a particular type (ints, floats, datetimes, strings etc) and each column can be treated as a Pandas Series.\n",
    "\n",
    "Above, we constructed data manually using lists and dicts. For the rest of this workshop, we will work with real data. DataFrames are the natural type to use when reading in tabular data from, for instance, CSV files or Excel files.\n",
    "\n",
    "Here's example to read in some small datasets which we can use for demo purposes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "iris = pd.read_csv('iris.csv')\n",
    "cars = pd.read_csv('mtcars.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data we'll use for this workshop is from the [Long Term Evolution Experiment (LTEE)](http://myxo.css.msu.edu/ecoli/). This experiment has been running for over 30 years and over 50,000 E. coli generations, and is still ongoing. Twelve separate populations of E. coli have been propagated for the life of the experiment. Every 500 generations, each population is cloned and stored, allowing researchers to study evolutionary behaviour over the long term, and to \"rewind and replay\" alternate evolutionary trajectories by propagating from an earlier clone. \n",
    "\n",
    "Several interesting events have occurred during the experiment. Some populations have spontaneously developed hypermutator phenotypes. In addition, around generation 31,000 one population, Ara-3, spontaneously developed a rare and novel Cit+ mutation, giving it the ability to metabolise citrate in the substrate.\n",
    "\n",
    "There have been many publications from this experiment. A handful:\n",
    "\n",
    "- [Blount et al 2008: Historical contingency and the evolution of a key innovation in an experimental population of Escherichia coli](https://www.pnas.org/content/105/23/7899) - on the spontaneous development of citrate metabolisation and on potentiating mutations\n",
    "- [Tenaillon et al 2016: Tempo and mode of genome evolution in a 50,000-generation experiment](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878) - various investigations by sequencing and variant-calling over 50,000 generations of clones, including discussion of hypermutator phenotypes and genetic drift vs natural selection.\n",
    "\n",
    "Sequence data from clones is available, but for this workshop we'll just be using some published tabular data.\n",
    "\n",
    "A version of this dataset is also used by the [Data Carpentry lessons on Genomics](https://datacarpentry.org/genomics-workshop/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the repository you'll find the files:\n",
    "\n",
    "- ltee_sampleruns.csv : sample and sequencing run metadata for the E. coli clones\n",
    "- ltee_mutations.csv : analysis output from variant calling on the E. coli clones\n",
    "- ltee_relative_fitness.tsv : relative fitness values for each population and generation up to generation 10,000\n",
    "- ltee_cell_size.tsv : cell sizes for each population and generation up to 10,000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's read in the sample and run metadata. Pandas has functions for reading in many data types. Try looking at the documentation for `read_csv()` by running `help(pd.read_csv)` or `pd.read_csv?`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleruns = pd.read_csv('ltee_sampleruns.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(264, 11)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The number of rows and columns\n",
    "sampleruns.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584408</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>70.241277</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584409</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>41.136778</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584410</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>33.226052</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584411</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>34.589932</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1062A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584438</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>40.342099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Population  Generation Strain ID Mutator   Reference   Facility   Accession  \\\n",
       "0      Ara+1         500   REL768A    None  This study  Genoscope  SRR2584408   \n",
       "1      Ara+1         500   REL768B    None  This study  Genoscope  SRR2584409   \n",
       "2      Ara+1        1000   REL958A    None  This study  Genoscope  SRR2584410   \n",
       "3      Ara+1        1000   REL958B    None  This study  Genoscope  SRR2584411   \n",
       "4      Ara+1        1500  REL1062A    None  This study  Genoscope  SRR2584438   \n",
       "\n",
       "  Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "0    single           36         70.241277            NaN  \n",
       "1    single           36         41.136778            NaN  \n",
       "2    single           36         33.226052            NaN  \n",
       "3    single           36         34.589932            NaN  \n",
       "4    single           36         40.342099            NaN  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The first few rows\n",
    "sampleruns.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Population           object\n",
       "Generation            int64\n",
       "Strain ID            object\n",
       "Mutator              object\n",
       "Reference            object\n",
       "Facility             object\n",
       "Accession            object\n",
       "Read Type            object\n",
       "Read Length           int64\n",
       "Sequencing Depth    float64\n",
       "Analysis Notes       object\n",
       "dtype: object"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Column data types\n",
    "sampleruns.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Population', 'Generation', 'Strain ID', 'Mutator', 'Reference',\n",
       "       'Facility', 'Accession', 'Read Type', 'Read Length', 'Sequencing Depth',\n",
       "       'Analysis Notes'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Column names\n",
    "sampleruns.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RangeIndex(start=0, stop=264, step=1)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Index (i.e. row names)\n",
    "# In this case we didn't provide an index and rows have simply been numbered for us by Pandas\n",
    "sampleruns.index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise\n",
    "\n",
    "1. Use `pd.read_csv()` to read the file `ltee_mutations.csv` into a variable called `mutations`.\n",
    "2. Check the column headings and the number of rows in this dataset, and have a look at the first few rows. Compare the size of the dataset and the variables to `sampleruns`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations = pd.read_csv('ltee_mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Total Mutations</th>\n",
       "      <th>Base Substitutions</th>\n",
       "      <th>IS Element Insertions</th>\n",
       "      <th>Small Indels</th>\n",
       "      <th>Large Deletions</th>\n",
       "      <th>Large Duplications</th>\n",
       "      <th>Total Deleted Base Pairs</th>\n",
       "      <th>Total Inserted Base Pairs</th>\n",
       "      <th>Estimated Final Genome Size</th>\n",
       "      <th>Synonymous Base Substitutions</th>\n",
       "      <th>Nonsynonymous Base Substitutions</th>\n",
       "      <th>Nonsense Base Substitutions</th>\n",
       "      <th>Pseudogene Base Substitutions</th>\n",
       "      <th>Noncoding RNA Base Substitutions</th>\n",
       "      <th>Intergenic Base Substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>REL768A</td>\n",
       "      <td>500</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6244</td>\n",
       "      <td>0</td>\n",
       "      <td>4623568</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>REL768B</td>\n",
       "      <td>500</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4772</td>\n",
       "      <td>1349</td>\n",
       "      <td>4626389</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>REL958A</td>\n",
       "      <td>1000</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2544</td>\n",
       "      <td>0</td>\n",
       "      <td>4627268</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>REL958B</td>\n",
       "      <td>1000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1446</td>\n",
       "      <td>4631258</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>REL1062A</td>\n",
       "      <td>1500</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>6305</td>\n",
       "      <td>0</td>\n",
       "      <td>4623507</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Population Strain ID  Generation  Total Mutations  Base Substitutions  \\\n",
       "0      Ara+1   REL768A         500                4                   3   \n",
       "1      Ara+1   REL768B         500                3                   1   \n",
       "2      Ara+1   REL958A        1000                4                   2   \n",
       "3      Ara+1   REL958B        1000                2                   1   \n",
       "4      Ara+1  REL1062A        1500                7                   5   \n",
       "\n",
       "   IS Element Insertions  Small Indels  Large Deletions  Large Duplications  \\\n",
       "0                      0             0                1                   0   \n",
       "1                      1             0                1                   0   \n",
       "2                      0             1                1                   0   \n",
       "3                      1             0                0                   0   \n",
       "4                      0             0                2                   0   \n",
       "\n",
       "   Total Deleted Base Pairs  Total Inserted Base Pairs  \\\n",
       "0                      6244                          0   \n",
       "1                      4772                       1349   \n",
       "2                      2544                          0   \n",
       "3                         0                       1446   \n",
       "4                      6305                          0   \n",
       "\n",
       "   Estimated Final Genome Size  Synonymous Base Substitutions  \\\n",
       "0                      4623568                              0   \n",
       "1                      4626389                              0   \n",
       "2                      4627268                              0   \n",
       "3                      4631258                              0   \n",
       "4                      4623507                              0   \n",
       "\n",
       "   Nonsynonymous Base Substitutions  Nonsense Base Substitutions  \\\n",
       "0                                 2                            0   \n",
       "1                                 1                            0   \n",
       "2                                 2                            0   \n",
       "3                                 1                            0   \n",
       "4                                 4                            0   \n",
       "\n",
       "   Pseudogene Base Substitutions  Noncoding RNA Base Substitutions  \\\n",
       "0                              0                                 0   \n",
       "1                              0                                 0   \n",
       "2                              0                                 0   \n",
       "3                              0                                 0   \n",
       "4                              0                                 0   \n",
       "\n",
       "   Intergenic Base Substitutions  \n",
       "0                              1  \n",
       "1                              0  \n",
       "2                              0  \n",
       "3                              0  \n",
       "4                              1  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing and slicing "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can extract a column from the DataFrame by indexing with square brackets, e.g. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracting a column\n",
    "sampleruns['Strain ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Firstly, let's set our index (row names) to something more meaningful to make it easier to see what's going on. The Strain ID uniquely identifies each sample, so it is probably a sensible index. We can use `sampleruns.set_index()`, or we can assign to the index directly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>REL768A</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584408</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>70.241277</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL768B</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584409</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>41.136778</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958A</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584410</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>33.226052</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958B</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584411</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>34.589932</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL1062A</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1062A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584438</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>40.342099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Population  Generation Strain ID Mutator   Reference   Facility  \\\n",
       "Strain ID                                                                   \n",
       "REL768A        Ara+1         500   REL768A    None  This study  Genoscope   \n",
       "REL768B        Ara+1         500   REL768B    None  This study  Genoscope   \n",
       "REL958A        Ara+1        1000   REL958A    None  This study  Genoscope   \n",
       "REL958B        Ara+1        1000   REL958B    None  This study  Genoscope   \n",
       "REL1062A       Ara+1        1500  REL1062A    None  This study  Genoscope   \n",
       "\n",
       "            Accession Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "Strain ID                                                                      \n",
       "REL768A    SRR2584408    single           36         70.241277            NaN  \n",
       "REL768B    SRR2584409    single           36         41.136778            NaN  \n",
       "REL958A    SRR2584410    single           36         33.226052            NaN  \n",
       "REL958B    SRR2584411    single           36         34.589932            NaN  \n",
       "REL1062A   SRR2584438    single           36         40.342099            NaN  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns.index = sampleruns['Strain ID']\n",
    "sampleruns.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The two most important ways to extract data from a DataFrame are `loc` and `iloc`. `loc` uses the index and the column names; `iloc` uses the row and column numbers, counting from zero. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SRR2584409'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns.loc['REL768B', 'Accession']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70.24127718360918"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Row 0, column 9\n",
    "sampleruns.iloc[0,9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Strain ID\n",
       "REL768B    41.136778\n",
       "REL958A    33.226052\n",
       "REL958B    34.589932\n",
       "Name: Sequencing Depth, dtype: float64"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rows 1-3, column 9\n",
    "sampleruns.iloc[1:4, 9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Population               Ara+1\n",
       "Generation                 500\n",
       "Strain ID              REL768B\n",
       "Mutator                   None\n",
       "Reference           This study\n",
       "Facility             Genoscope\n",
       "Accession           SRR2584409\n",
       "Read Type               single\n",
       "Read Length                 36\n",
       "Sequencing Depth       41.1368\n",
       "Analysis Notes             NaN\n",
       "Name: REL768B, dtype: object"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All columns\n",
    "sampleruns.loc['REL768B', :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>REL768A</th>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958A</th>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Read Type  Read Length\n",
       "Strain ID                       \n",
       "REL768A      single           36\n",
       "REL958A      single           36"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lists let us specify any set of rows and columns, in any order\n",
    "sampleruns.loc[['REL768A','REL958A'], ['Read Type', 'Read Length']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sepal_length</th>\n",
       "      <th>petal_length</th>\n",
       "      <th>petal_width</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5.1</td>\n",
       "      <td>1.4</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4.9</td>\n",
       "      <td>1.4</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.7</td>\n",
       "      <td>1.3</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.6</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.4</td>\n",
       "      <td>1.7</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sepal_length  petal_length  petal_width\n",
       "0           5.1           1.4          0.2\n",
       "1           4.9           1.4          0.2\n",
       "2           4.7           1.3          0.2\n",
       "3           4.6           1.5          0.2\n",
       "4           5.0           1.4          0.2\n",
       "5           5.4           1.7          0.4"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We can also use True/False values to perform boolean indexing. \n",
    "# Pandas will return the rows/columns matching the True values we pass in.\n",
    "# This will be useful later for filtering data\n",
    "iris.loc[0:5, [True, False, True, True, False]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise\n",
    "\n",
    "1. Set the index of `mutations` to be the same as the \"Strain ID\" column.\n",
    "2. Extract the population, generation, and number of total mutations for strain REL11345"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations.index = mutations['Strain ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Population         Ara+3\n",
       "Generation         50000\n",
       "Total Mutations     1729\n",
       "Name: REL11345, dtype: object"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.loc['REL11345', ['Population','Generation', 'Total Mutations']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manipulating Series"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A single column of a DataFrame is a Series object. Series have a data type:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Sequencing Depth'].dtype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Like a DataFrame, a Series has an index. In this case we got our Series from a column of a DataFrame, so it will have the same index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['REL768A', 'REL768B', 'REL958A', 'REL958B', 'REL1062A', 'REL1062B',\n",
       "       'REL1158A', 'REL1158C', 'REL2173A', 'REL2173B',\n",
       "       ...\n",
       "       'REL7182A', 'REL7182B', 'REL8598A', 'REL8598B', 'REL10447', 'REL10448',\n",
       "       'REL11005', 'REL11006', 'REL11389', 'REL11390'],\n",
       "      dtype='object', name='Strain ID', length=264)"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Sequencing Depth'].index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are several convenience functions defined on Series. For instance, we can find the average sequencing depth:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "115.77451401512721"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Sequencing Depth'].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly for numeric variables we have, for instance, `.min()` and `.max()`, `.median()`, `std()`, and `sum()`.\n",
    "\n",
    "`.describe()` is a convenience function for getting several summary statistics at once:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     264.000000\n",
       "mean      115.774514\n",
       "std       149.345576\n",
       "min        27.803535\n",
       "25%        60.845261\n",
       "50%        79.629418\n",
       "75%       106.138199\n",
       "max      1026.249751\n",
       "Name: Sequencing Depth, dtype: float64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Sequencing Depth'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For non-numeric variable types such as strings and categoricals, we may want to look at the unique values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['single', 'paired'], dtype=object)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Read Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "paired    134\n",
       "single    130\n",
       "Name: Read Type, dtype: int64"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Read Type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas makes use of numpy vectorisation, meaning we can do operations on Series with simple syntax, and it will be efficient to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Strain ID\n",
       "REL768A     70.241277\n",
       "REL768B     41.136778\n",
       "REL958A     33.226052\n",
       "REL958B     34.589932\n",
       "REL1062A    40.342099\n",
       "Name: Sequencing Depth, dtype: float64"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: use mutations unless exercise\n",
    "sampleruns['Sequencing Depth'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Strain ID\n",
       "REL768A     75.241277\n",
       "REL768B     46.136778\n",
       "REL958A     38.226052\n",
       "REL958B     39.589932\n",
       "REL1062A    45.342099\n",
       "Name: Sequencing Depth, dtype: float64"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns['Sequencing Depth'].head() + 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You might have noticed that the `Analysis Notes` column contains a lot of NaN's. This means \"not a number\" and represents a missing value - i.e. these cells are empty. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>REL768A</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584408</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>70.241277</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL768B</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584409</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>41.136778</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958A</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584410</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>33.226052</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958B</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584411</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>34.589932</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL1062A</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1062A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584438</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>40.342099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Population  Generation Strain ID Mutator   Reference   Facility  \\\n",
       "Strain ID                                                                   \n",
       "REL768A        Ara+1         500   REL768A    None  This study  Genoscope   \n",
       "REL768B        Ara+1         500   REL768B    None  This study  Genoscope   \n",
       "REL958A        Ara+1        1000   REL958A    None  This study  Genoscope   \n",
       "REL958B        Ara+1        1000   REL958B    None  This study  Genoscope   \n",
       "REL1062A       Ara+1        1500  REL1062A    None  This study  Genoscope   \n",
       "\n",
       "            Accession Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "Strain ID                                                                      \n",
       "REL768A    SRR2584408    single           36         70.241277            NaN  \n",
       "REL768B    SRR2584409    single           36         41.136778            NaN  \n",
       "REL958A    SRR2584410    single           36         33.226052            NaN  \n",
       "REL958B    SRR2584411    single           36         34.589932            NaN  \n",
       "REL1062A   SRR2584438    single           36         40.342099            NaN  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can check which values are missing with `.isnull()`. This converts every value in the DataFrame (or Series) into a boolean True/False value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>REL768A</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL768B</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958A</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL958B</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL1062A</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Population  Generation  Strain ID  Mutator  Reference  Facility  \\\n",
       "Strain ID                                                                    \n",
       "REL768A         False       False      False    False      False     False   \n",
       "REL768B         False       False      False    False      False     False   \n",
       "REL958A         False       False      False    False      False     False   \n",
       "REL958B         False       False      False    False      False     False   \n",
       "REL1062A        False       False      False    False      False     False   \n",
       "\n",
       "           Accession  Read Type  Read Length  Sequencing Depth  Analysis Notes  \n",
       "Strain ID                                                                       \n",
       "REL768A        False      False        False             False            True  \n",
       "REL768B        False      False        False             False            True  \n",
       "REL958A        False      False        False             False            True  \n",
       "REL958B        False      False        False             False            True  \n",
       "REL1062A       False      False        False             False            True  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns.isnull().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adding up booleans will treat `True` as `1` and `False` as `0`. A common approach is to use `sum()` to count how many `True` values there are. So we can count missing values like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Population            0\n",
       "Generation            0\n",
       "Strain ID             0\n",
       "Mutator               0\n",
       "Reference             0\n",
       "Facility              0\n",
       "Accession             0\n",
       "Read Type             0\n",
       "Read Length           0\n",
       "Sequencing Depth      0\n",
       "Analysis Notes      251\n",
       "dtype: int64"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleruns.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`sampleruns` had 264 rows, so it looks like there are a few non-empty note cells. We could count this explicitly by taking the logical `not` of our True/False values, i.e. adding up cells where `isnull()` is `False`. For manipulating array-like data, we can't use the `not`, `and` and `or` boolean operators. Instead we need to use the bitwise operators `~`, `&`, and `|`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Population          264\n",
       "Generation          264\n",
       "Strain ID           264\n",
       "Mutator             264\n",
       "Reference           264\n",
       "Facility            264\n",
       "Accession           264\n",
       "Read Type           264\n",
       "Read Length         264\n",
       "Sequencing Depth    264\n",
       "Analysis Notes       13\n",
       "dtype: int64"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(~sampleruns.isnull()).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that Pandas summed each column. We can use `sum(axis=1)` to override this default and sum each row instead."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sorting and filtering "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can sort on a field, or list of fields, with `.sort_values()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11368</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584822</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>100.551741</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>10000</td>\n",
       "      <td>REL4537B</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584783</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>33.303734</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ara+2</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL959A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584583</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>60.081156</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>20000</td>\n",
       "      <td>REL8598B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584978</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>91.769745</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11337</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584889</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>78.717886</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>500</td>\n",
       "      <td>REL765A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584867</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>62.015488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10448</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2589077</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>97.147508</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10433</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584820</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>55.810155</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>5000</td>\n",
       "      <td>REL2176B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2588991</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>135.749446</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>500</td>\n",
       "      <td>REL771B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584698</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>130.838758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11333</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584849</td>\n",
       "      <td>single</td>\n",
       "      <td>51</td>\n",
       "      <td>770.898034</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1070B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584858</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>102.377792</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>15000</td>\n",
       "      <td>REL7182A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584942</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>91.040975</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10938</td>\n",
       "      <td>PM</td>\n",
       "      <td>B09</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR030258</td>\n",
       "      <td>paired</td>\n",
       "      <td>36</td>\n",
       "      <td>60.706100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10983</td>\n",
       "      <td>None</td>\n",
       "      <td>W13</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2086150</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>64.871361</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Population  Generation Strain ID Mutator   Reference   Facility  \\\n",
       "109      Ara+5       50000  REL11368    None  This study   IntraGen   \n",
       "165      Ara-2       10000  REL4537B      PM  This study   MSU RTSF   \n",
       "24       Ara+2        1000   REL959A    None  This study  Genoscope   \n",
       "257      Ara-6       20000  REL8598B    None  This study   IntraGen   \n",
       "219      Ara-4       50000  REL11337      PM  This study   MSU RTSF   \n",
       "198      Ara-4         500   REL765A    None  This study   MSU RTSF   \n",
       "259      Ara-6       30000  REL10448    None  This study   IntraGen   \n",
       "105      Ara+5       30000  REL10433    None  This study  Genoscope   \n",
       "75       Ara+4        5000  REL2176B    None  This study   IntraGen   \n",
       "67       Ara+4         500   REL771B    None  This study   IntraGen   \n",
       "174      Ara-2       50000  REL11333      PM  This study  Genoscope   \n",
       "181      Ara-3        1500  REL1070B    None  This study   IntraGen   \n",
       "254      Ara-6       15000  REL7182A    None  This study   IntraGen   \n",
       "150      Ara-1       40000  REL10938      PM         B09   MSU RTSF   \n",
       "107      Ara+5       40000  REL10983    None         W13  Genoscope   \n",
       "\n",
       "      Accession Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "109  SRR2584822    paired          101        100.551741            NaN  \n",
       "165  SRR2584783    paired          150         33.303734            NaN  \n",
       "24   SRR2584583    single           36         60.081156            NaN  \n",
       "257  SRR2584978    paired          101         91.769745            NaN  \n",
       "219  SRR2584889    paired          150         78.717886            NaN  \n",
       "198  SRR2584867    paired          150         62.015488            NaN  \n",
       "259  SRR2589077    paired          101         97.147508            NaN  \n",
       "105  SRR2584820    single           36         55.810155            NaN  \n",
       "75   SRR2588991    paired          101        135.749446            NaN  \n",
       "67   SRR2584698    paired          101        130.838758            NaN  \n",
       "174  SRR2584849    single           51        770.898034            NaN  \n",
       "181  SRR2584858    paired          101        102.377792            NaN  \n",
       "254  SRR2584942    paired          101         91.040975            NaN  \n",
       "150   SRR030258    paired           36         60.706100            NaN  \n",
       "107  SRR2086150    single           36         64.871361            NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get a random example subset\n",
    "subset = sampleruns.sample(15)\n",
    "subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>500</td>\n",
       "      <td>REL765A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584867</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>62.015488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>500</td>\n",
       "      <td>REL771B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584698</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>130.838758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ara+2</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL959A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584583</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>60.081156</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1070B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584858</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>102.377792</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>5000</td>\n",
       "      <td>REL2176B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2588991</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>135.749446</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>10000</td>\n",
       "      <td>REL4537B</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584783</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>33.303734</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>15000</td>\n",
       "      <td>REL7182A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584942</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>91.040975</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>20000</td>\n",
       "      <td>REL8598B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584978</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>91.769745</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10448</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2589077</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>97.147508</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10433</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584820</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>55.810155</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10938</td>\n",
       "      <td>PM</td>\n",
       "      <td>B09</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR030258</td>\n",
       "      <td>paired</td>\n",
       "      <td>36</td>\n",
       "      <td>60.706100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10983</td>\n",
       "      <td>None</td>\n",
       "      <td>W13</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2086150</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>64.871361</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11368</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584822</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>100.551741</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11337</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584889</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>78.717886</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11333</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584849</td>\n",
       "      <td>single</td>\n",
       "      <td>51</td>\n",
       "      <td>770.898034</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Population  Generation Strain ID Mutator   Reference   Facility  \\\n",
       "198      Ara-4         500   REL765A    None  This study   MSU RTSF   \n",
       "67       Ara+4         500   REL771B    None  This study   IntraGen   \n",
       "24       Ara+2        1000   REL959A    None  This study  Genoscope   \n",
       "181      Ara-3        1500  REL1070B    None  This study   IntraGen   \n",
       "75       Ara+4        5000  REL2176B    None  This study   IntraGen   \n",
       "165      Ara-2       10000  REL4537B      PM  This study   MSU RTSF   \n",
       "254      Ara-6       15000  REL7182A    None  This study   IntraGen   \n",
       "257      Ara-6       20000  REL8598B    None  This study   IntraGen   \n",
       "259      Ara-6       30000  REL10448    None  This study   IntraGen   \n",
       "105      Ara+5       30000  REL10433    None  This study  Genoscope   \n",
       "150      Ara-1       40000  REL10938      PM         B09   MSU RTSF   \n",
       "107      Ara+5       40000  REL10983    None         W13  Genoscope   \n",
       "109      Ara+5       50000  REL11368    None  This study   IntraGen   \n",
       "219      Ara-4       50000  REL11337      PM  This study   MSU RTSF   \n",
       "174      Ara-2       50000  REL11333      PM  This study  Genoscope   \n",
       "\n",
       "      Accession Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "198  SRR2584867    paired          150         62.015488            NaN  \n",
       "67   SRR2584698    paired          101        130.838758            NaN  \n",
       "24   SRR2584583    single           36         60.081156            NaN  \n",
       "181  SRR2584858    paired          101        102.377792            NaN  \n",
       "75   SRR2588991    paired          101        135.749446            NaN  \n",
       "165  SRR2584783    paired          150         33.303734            NaN  \n",
       "254  SRR2584942    paired          101         91.040975            NaN  \n",
       "257  SRR2584978    paired          101         91.769745            NaN  \n",
       "259  SRR2589077    paired          101         97.147508            NaN  \n",
       "105  SRR2584820    single           36         55.810155            NaN  \n",
       "150   SRR030258    paired           36         60.706100            NaN  \n",
       "107  SRR2086150    single           36         64.871361            NaN  \n",
       "109  SRR2584822    paired          101        100.551741            NaN  \n",
       "219  SRR2584889    paired          150         78.717886            NaN  \n",
       "174  SRR2584849    single           51        770.898034            NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset.sort_values('Generation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ara+2</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL959A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584583</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>60.081156</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>500</td>\n",
       "      <td>REL771B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584698</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>130.838758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>5000</td>\n",
       "      <td>REL2176B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2588991</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>135.749446</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10433</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584820</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>55.810155</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10983</td>\n",
       "      <td>None</td>\n",
       "      <td>W13</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2086150</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>64.871361</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11368</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584822</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>100.551741</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10938</td>\n",
       "      <td>PM</td>\n",
       "      <td>B09</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR030258</td>\n",
       "      <td>paired</td>\n",
       "      <td>36</td>\n",
       "      <td>60.706100</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>10000</td>\n",
       "      <td>REL4537B</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584783</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>33.303734</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11333</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584849</td>\n",
       "      <td>single</td>\n",
       "      <td>51</td>\n",
       "      <td>770.898034</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1070B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584858</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>102.377792</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>500</td>\n",
       "      <td>REL765A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584867</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>62.015488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11337</td>\n",
       "      <td>PM</td>\n",
       "      <td>This study</td>\n",
       "      <td>MSU RTSF</td>\n",
       "      <td>SRR2584889</td>\n",
       "      <td>paired</td>\n",
       "      <td>150</td>\n",
       "      <td>78.717886</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>15000</td>\n",
       "      <td>REL7182A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584942</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>91.040975</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>20000</td>\n",
       "      <td>REL8598B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584978</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>91.769745</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10448</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2589077</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>97.147508</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Population  Generation Strain ID Mutator   Reference   Facility  \\\n",
       "24       Ara+2        1000   REL959A    None  This study  Genoscope   \n",
       "67       Ara+4         500   REL771B    None  This study   IntraGen   \n",
       "75       Ara+4        5000  REL2176B    None  This study   IntraGen   \n",
       "105      Ara+5       30000  REL10433    None  This study  Genoscope   \n",
       "107      Ara+5       40000  REL10983    None         W13  Genoscope   \n",
       "109      Ara+5       50000  REL11368    None  This study   IntraGen   \n",
       "150      Ara-1       40000  REL10938      PM         B09   MSU RTSF   \n",
       "165      Ara-2       10000  REL4537B      PM  This study   MSU RTSF   \n",
       "174      Ara-2       50000  REL11333      PM  This study  Genoscope   \n",
       "181      Ara-3        1500  REL1070B    None  This study   IntraGen   \n",
       "198      Ara-4         500   REL765A    None  This study   MSU RTSF   \n",
       "219      Ara-4       50000  REL11337      PM  This study   MSU RTSF   \n",
       "254      Ara-6       15000  REL7182A    None  This study   IntraGen   \n",
       "257      Ara-6       20000  REL8598B    None  This study   IntraGen   \n",
       "259      Ara-6       30000  REL10448    None  This study   IntraGen   \n",
       "\n",
       "      Accession Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "24   SRR2584583    single           36         60.081156            NaN  \n",
       "67   SRR2584698    paired          101        130.838758            NaN  \n",
       "75   SRR2588991    paired          101        135.749446            NaN  \n",
       "105  SRR2584820    single           36         55.810155            NaN  \n",
       "107  SRR2086150    single           36         64.871361            NaN  \n",
       "109  SRR2584822    paired          101        100.551741            NaN  \n",
       "150   SRR030258    paired           36         60.706100            NaN  \n",
       "165  SRR2584783    paired          150         33.303734            NaN  \n",
       "174  SRR2584849    single           51        770.898034            NaN  \n",
       "181  SRR2584858    paired          101        102.377792            NaN  \n",
       "198  SRR2584867    paired          150         62.015488            NaN  \n",
       "219  SRR2584889    paired          150         78.717886            NaN  \n",
       "254  SRR2584942    paired          101         91.040975            NaN  \n",
       "257  SRR2584978    paired          101         91.769745            NaN  \n",
       "259  SRR2589077    paired          101         97.147508            NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset.sort_values(['Population','Generation'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can filter datasets using boolean indexing. This means that if we use a logical expression produce a boolean Series with a logical expression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "109     True\n",
       "165    False\n",
       "24     False\n",
       "257    False\n",
       "219    False\n",
       "198    False\n",
       "259    False\n",
       "105     True\n",
       "75     False\n",
       "67     False\n",
       "174    False\n",
       "181    False\n",
       "254    False\n",
       "150    False\n",
       "107     True\n",
       "Name: Population, dtype: bool"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset['Population'] == 'Ara+5'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we can then select out only the rows (or sometimes columns) where that logical expression is True:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>Analysis Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>50000</td>\n",
       "      <td>REL11368</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>IntraGen</td>\n",
       "      <td>SRR2584822</td>\n",
       "      <td>paired</td>\n",
       "      <td>101</td>\n",
       "      <td>100.551741</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>30000</td>\n",
       "      <td>REL10433</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584820</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>55.810155</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>40000</td>\n",
       "      <td>REL10983</td>\n",
       "      <td>None</td>\n",
       "      <td>W13</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2086150</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>64.871361</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Population  Generation Strain ID Mutator   Reference   Facility  \\\n",
       "109      Ara+5       50000  REL11368    None  This study   IntraGen   \n",
       "105      Ara+5       30000  REL10433    None  This study  Genoscope   \n",
       "107      Ara+5       40000  REL10983    None         W13  Genoscope   \n",
       "\n",
       "      Accession Read Type  Read Length  Sequencing Depth Analysis Notes  \n",
       "109  SRR2584822    paired          101        100.551741            NaN  \n",
       "105  SRR2584820    single           36         55.810155            NaN  \n",
       "107  SRR2086150    single           36         64.871361            NaN  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset.loc[subset['Population'] == 'Ara+5', :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise \n",
    "\n",
    "1. Filter the `mutations` dataset to extract only samples with more than 1500 total mutations. \n",
    "2. Sort the resulting data by the number of \"Small Indels\". Have a look at the resulting Population and Generation columns.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Total Mutations</th>\n",
       "      <th>Base Substitutions</th>\n",
       "      <th>IS Element Insertions</th>\n",
       "      <th>Small Indels</th>\n",
       "      <th>Large Deletions</th>\n",
       "      <th>Large Duplications</th>\n",
       "      <th>Total Deleted Base Pairs</th>\n",
       "      <th>Total Inserted Base Pairs</th>\n",
       "      <th>Estimated Final Genome Size</th>\n",
       "      <th>Synonymous Base Substitutions</th>\n",
       "      <th>Nonsynonymous Base Substitutions</th>\n",
       "      <th>Nonsense Base Substitutions</th>\n",
       "      <th>Pseudogene Base Substitutions</th>\n",
       "      <th>Noncoding RNA Base Substitutions</th>\n",
       "      <th>Intergenic Base Substitutions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>REL10417</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>REL10417</td>\n",
       "      <td>30000</td>\n",
       "      <td>1837</td>\n",
       "      <td>1827</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>43480</td>\n",
       "      <td>4352</td>\n",
       "      <td>4590684</td>\n",
       "      <td>214</td>\n",
       "      <td>1316</td>\n",
       "      <td>32</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL10418</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>REL10418</td>\n",
       "      <td>30000</td>\n",
       "      <td>1883</td>\n",
       "      <td>1874</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>67635</td>\n",
       "      <td>2798</td>\n",
       "      <td>4564975</td>\n",
       "      <td>211</td>\n",
       "      <td>1360</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL10985</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>REL10985</td>\n",
       "      <td>40000</td>\n",
       "      <td>2185</td>\n",
       "      <td>2172</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>76300</td>\n",
       "      <td>3576</td>\n",
       "      <td>4557088</td>\n",
       "      <td>268</td>\n",
       "      <td>1561</td>\n",
       "      <td>35</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL11370</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>REL11370</td>\n",
       "      <td>50000</td>\n",
       "      <td>2557</td>\n",
       "      <td>2548</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>73042</td>\n",
       "      <td>2799</td>\n",
       "      <td>4559569</td>\n",
       "      <td>292</td>\n",
       "      <td>1818</td>\n",
       "      <td>42</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL11371</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>REL11371</td>\n",
       "      <td>50000</td>\n",
       "      <td>2293</td>\n",
       "      <td>2276</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>62344</td>\n",
       "      <td>5902</td>\n",
       "      <td>4573370</td>\n",
       "      <td>287</td>\n",
       "      <td>1629</td>\n",
       "      <td>38</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL10986</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>REL10986</td>\n",
       "      <td>40000</td>\n",
       "      <td>2021</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>43481</td>\n",
       "      <td>5693</td>\n",
       "      <td>4592024</td>\n",
       "      <td>227</td>\n",
       "      <td>1443</td>\n",
       "      <td>38</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL10954</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>REL10954</td>\n",
       "      <td>40000</td>\n",
       "      <td>1537</td>\n",
       "      <td>1270</td>\n",
       "      <td>4</td>\n",
       "      <td>259</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>70598</td>\n",
       "      <td>5758</td>\n",
       "      <td>4564972</td>\n",
       "      <td>390</td>\n",
       "      <td>709</td>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL10953</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>REL10953</td>\n",
       "      <td>40000</td>\n",
       "      <td>1532</td>\n",
       "      <td>1262</td>\n",
       "      <td>4</td>\n",
       "      <td>261</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>92750</td>\n",
       "      <td>5757</td>\n",
       "      <td>4542819</td>\n",
       "      <td>390</td>\n",
       "      <td>706</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL11346</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>REL11346</td>\n",
       "      <td>50000</td>\n",
       "      <td>1640</td>\n",
       "      <td>1361</td>\n",
       "      <td>5</td>\n",
       "      <td>269</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>92764</td>\n",
       "      <td>5873</td>\n",
       "      <td>4542921</td>\n",
       "      <td>408</td>\n",
       "      <td>763</td>\n",
       "      <td>26</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REL11345</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>REL11345</td>\n",
       "      <td>50000</td>\n",
       "      <td>1729</td>\n",
       "      <td>1448</td>\n",
       "      <td>4</td>\n",
       "      <td>271</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>100311</td>\n",
       "      <td>5766</td>\n",
       "      <td>4535267</td>\n",
       "      <td>451</td>\n",
       "      <td>810</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Population Strain ID  Generation  Total Mutations  \\\n",
       "Strain ID                                                     \n",
       "REL10417       Ara+6  REL10417       30000             1837   \n",
       "REL10418       Ara+6  REL10418       30000             1883   \n",
       "REL10985       Ara+6  REL10985       40000             2185   \n",
       "REL11370       Ara+6  REL11370       50000             2557   \n",
       "REL11371       Ara+6  REL11371       50000             2293   \n",
       "REL10986       Ara+6  REL10986       40000             2021   \n",
       "REL10954       Ara+3  REL10954       40000             1537   \n",
       "REL10953       Ara+3  REL10953       40000             1532   \n",
       "REL11346       Ara+3  REL11346       50000             1640   \n",
       "REL11345       Ara+3  REL11345       50000             1729   \n",
       "\n",
       "           Base Substitutions  IS Element Insertions  Small Indels  \\\n",
       "Strain ID                                                            \n",
       "REL10417                 1827                      4             2   \n",
       "REL10418                 1874                      2             2   \n",
       "REL10985                 2172                      3             2   \n",
       "REL11370                 2548                      2             2   \n",
       "REL11371                 2276                      6             3   \n",
       "REL10986                 2009                      4             4   \n",
       "REL10954                 1270                      4           259   \n",
       "REL10953                 1262                      4           261   \n",
       "REL11346                 1361                      5           269   \n",
       "REL11345                 1448                      4           271   \n",
       "\n",
       "           Large Deletions  Large Duplications  Total Deleted Base Pairs  \\\n",
       "Strain ID                                                                  \n",
       "REL10417                 4                   0                     43480   \n",
       "REL10418                 5                   0                     67635   \n",
       "REL10985                 8                   0                     76300   \n",
       "REL11370                 5                   0                     73042   \n",
       "REL11371                 8                   0                     62344   \n",
       "REL10986                 4                   0                     43481   \n",
       "REL10954                 4                   0                     70598   \n",
       "REL10953                 5                   0                     92750   \n",
       "REL11346                 5                   0                     92764   \n",
       "REL11345                 6                   0                    100311   \n",
       "\n",
       "           Total Inserted Base Pairs  Estimated Final Genome Size  \\\n",
       "Strain ID                                                           \n",
       "REL10417                        4352                      4590684   \n",
       "REL10418                        2798                      4564975   \n",
       "REL10985                        3576                      4557088   \n",
       "REL11370                        2799                      4559569   \n",
       "REL11371                        5902                      4573370   \n",
       "REL10986                        5693                      4592024   \n",
       "REL10954                        5758                      4564972   \n",
       "REL10953                        5757                      4542819   \n",
       "REL11346                        5873                      4542921   \n",
       "REL11345                        5766                      4535267   \n",
       "\n",
       "           Synonymous Base Substitutions  Nonsynonymous Base Substitutions  \\\n",
       "Strain ID                                                                    \n",
       "REL10417                             214                              1316   \n",
       "REL10418                             211                              1360   \n",
       "REL10985                             268                              1561   \n",
       "REL11370                             292                              1818   \n",
       "REL11371                             287                              1629   \n",
       "REL10986                             227                              1443   \n",
       "REL10954                             390                               709   \n",
       "REL10953                             390                               706   \n",
       "REL11346                             408                               763   \n",
       "REL11345                             451                               810   \n",
       "\n",
       "           Nonsense Base Substitutions  Pseudogene Base Substitutions  \\\n",
       "Strain ID                                                               \n",
       "REL10417                            32                             23   \n",
       "REL10418                            30                             22   \n",
       "REL10985                            35                             25   \n",
       "REL11370                            42                             39   \n",
       "REL11371                            38                             28   \n",
       "REL10986                            38                             29   \n",
       "REL10954                            24                             19   \n",
       "REL10953                            22                             17   \n",
       "REL11346                            26                             20   \n",
       "REL11345                            24                             20   \n",
       "\n",
       "           Noncoding RNA Base Substitutions  Intergenic Base Substitutions  \n",
       "Strain ID                                                                   \n",
       "REL10417                                  1                            241  \n",
       "REL10418                                  1                            250  \n",
       "REL10985                                  1                            282  \n",
       "REL11370                                  1                            356  \n",
       "REL11371                                  1                            293  \n",
       "REL10986                                  1                            271  \n",
       "REL10954                                  0                            128  \n",
       "REL10953                                  0                            127  \n",
       "REL11346                                  0                            144  \n",
       "REL11345                                  0                            143  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations[mutations['Total Mutations'] > 1500].sort_values('Small Indels')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interlude: tidy data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Link to \"untidy data\" discussion spreadsheet](https://docs.google.com/spreadsheets/d/1P94oEzgxNzlpvYiento53tZxJwaHYi8gpcdpDouu2jw/edit?usp=sharing)\n",
    "\n",
    "(Don't look too far into this spreadsheet before we get up to it, as it includes solutions which are a spoiler for the exercise.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reshaping data "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can reshape data using particular columns, with `melt` and `pivot` or `pivot_table`. We'll have a look at this below.\n",
    "\n",
    "We can also reshape data using the column names and index, with `stack` and `unstack`. This requires MultiIndexes, which we won't go into today."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are two tiny \"wide\" datasets based on our \"untidy\" housing data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-19T11:51:16.434295Z",
     "start_time": "2019-01-19T11:51:16.254462Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>property_id</th>\n",
       "      <th>bedrooms</th>\n",
       "      <th>price1</th>\n",
       "      <th>price2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>180000.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>430000.0</td>\n",
       "      <td>700000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>221900.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   property_id  bedrooms    price1    price2\n",
       "0   5631500400         2  180000.0       NaN\n",
       "1   6021501535         3  430000.0  700000.0\n",
       "2   7129300520         3  221900.0       NaN"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sales_wide1 = pd.read_csv('housing-data-wide1.csv')\n",
    "sales_wide1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-19T11:54:26.755381Z",
     "start_time": "2019-01-19T11:54:26.723723Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>property_id</th>\n",
       "      <th>bedrooms</th>\n",
       "      <th>date1</th>\n",
       "      <th>price1</th>\n",
       "      <th>date2</th>\n",
       "      <th>price2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>2015-02-25</td>\n",
       "      <td>180000.0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>2014-07-25</td>\n",
       "      <td>430000.0</td>\n",
       "      <td>2014-12-23</td>\n",
       "      <td>700000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>2014-10-13</td>\n",
       "      <td>221900.0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   property_id  bedrooms      date1    price1      date2    price2\n",
       "0   5631500400         2 2015-02-25  180000.0        NaT       NaN\n",
       "1   6021501535         3 2014-07-25  430000.0 2014-12-23  700000.0\n",
       "2   7129300520         3 2014-10-13  221900.0        NaT       NaN"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sales_wide2 = pd.read_csv('housing-data-wide2.csv', parse_dates=['date1','date2'])\n",
    "sales_wide2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How can we reshape these into tidy form? \n",
    "\n",
    "The Pandas `melt` function will do this. For instance:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>property_id</td>\n",
       "      <td>5.631500e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>property_id</td>\n",
       "      <td>6.021502e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>property_id</td>\n",
       "      <td>7.129301e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bedrooms</td>\n",
       "      <td>2.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bedrooms</td>\n",
       "      <td>3.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>bedrooms</td>\n",
       "      <td>3.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>price1</td>\n",
       "      <td>1.800000e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>price1</td>\n",
       "      <td>4.300000e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>price1</td>\n",
       "      <td>2.219000e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>price2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>price2</td>\n",
       "      <td>7.000000e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>price2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       variable         value\n",
       "0   property_id  5.631500e+09\n",
       "1   property_id  6.021502e+09\n",
       "2   property_id  7.129301e+09\n",
       "3      bedrooms  2.000000e+00\n",
       "4      bedrooms  3.000000e+00\n",
       "5      bedrooms  3.000000e+00\n",
       "6        price1  1.800000e+05\n",
       "7        price1  4.300000e+05\n",
       "8        price1  2.219000e+05\n",
       "9        price2           NaN\n",
       "10       price2  7.000000e+05\n",
       "11       price2           NaN"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sales_wide1.melt()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This has put every variable (i.e. every column) into the new `variable` column. This probably isn't what we want. It's only the price columns that are \"wide\", the other variables were fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>property_id</th>\n",
       "      <th>bedrooms</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>price1</td>\n",
       "      <td>180000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>price1</td>\n",
       "      <td>430000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>price1</td>\n",
       "      <td>221900.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>price2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>price2</td>\n",
       "      <td>700000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>price2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   property_id  bedrooms variable     value\n",
       "0   5631500400         2   price1  180000.0\n",
       "1   6021501535         3   price1  430000.0\n",
       "2   7129300520         3   price1  221900.0\n",
       "3   5631500400         2   price2       NaN\n",
       "4   6021501535         3   price2  700000.0\n",
       "5   7129300520         3   price2       NaN"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Retain columns property_id and bedrooms\n",
    "sales_wide1.melt(id_vars=['property_id','bedrooms'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is getting close to what we want. The `variable` column contains the original column names and tells us whether the price we're looking at was from the first or second sale (this may or may not be information we care about). The `value` column contains values in the melted columns, i.e. the actual price. \n",
    "\n",
    "Now we technically have long form and have eliminated the duplicated `price` variable; all prices are now in the `value` column. Notice that properties can now appear more than once in the table; conceptually, we have a row per sale rather than a row per property. \n",
    "\n",
    "We can tell `melt()` what to call the `variable` and `value` columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>property_id</th>\n",
       "      <th>bedrooms</th>\n",
       "      <th>sale_number</th>\n",
       "      <th>price</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>price1</td>\n",
       "      <td>180000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>price1</td>\n",
       "      <td>430000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>price1</td>\n",
       "      <td>221900.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>price2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>price2</td>\n",
       "      <td>700000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>price2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   property_id  bedrooms sale_number     price\n",
       "0   5631500400         2      price1  180000.0\n",
       "1   6021501535         3      price1  430000.0\n",
       "2   7129300520         3      price1  221900.0\n",
       "3   5631500400         2      price2       NaN\n",
       "4   6021501535         3      price2  700000.0\n",
       "5   7129300520         3      price2       NaN"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sales_tidy = sales_wide1.melt(id_vars=['property_id','bedrooms'], \n",
    "                 var_name='sale_number',\n",
    "                 value_name='price')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We still have empty extra rows where there was no sale in the original table, i.e. rows 3 and 5. We could use `dropna()` to get rid of these. A more generic approach would be to use filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>property_id</th>\n",
       "      <th>bedrooms</th>\n",
       "      <th>sale_number</th>\n",
       "      <th>price</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5631500400</td>\n",
       "      <td>2</td>\n",
       "      <td>price1</td>\n",
       "      <td>180000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>price1</td>\n",
       "      <td>430000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7129300520</td>\n",
       "      <td>3</td>\n",
       "      <td>price1</td>\n",
       "      <td>221900.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6021501535</td>\n",
       "      <td>3</td>\n",
       "      <td>price2</td>\n",
       "      <td>700000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   property_id  bedrooms sale_number     price\n",
       "0   5631500400         2      price1  180000.0\n",
       "1   6021501535         3      price1  430000.0\n",
       "2   7129300520         3      price1  221900.0\n",
       "4   6021501535         3      price2  700000.0"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sales = sales_wide1.melt(id_vars=['property_id','bedrooms'], \n",
    "                                     var_name='sale_number',\n",
    "                                     value_name='price')\n",
    "sales = sales[~sales['price'].isnull()]\n",
    "sales"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is pretty good! Now we could run commands like `sales[\"price\"].mean()` and get a sensible answer. We'll also be able to use the data easily to produce plots.\n",
    "\n",
    "If you want a challenge, think about how you could convert `sales_wide2` to tidy form - it's a fair bit harder.\n",
    "\n",
    "The inverse operation to `.melt()` is `.pivot()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sale_number</th>\n",
       "      <th>price1</th>\n",
       "      <th>price2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>property_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5631500400</th>\n",
       "      <td>180000.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6021501535</th>\n",
       "      <td>430000.0</td>\n",
       "      <td>700000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129300520</th>\n",
       "      <td>221900.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "sale_number    price1    price2\n",
       "property_id                    \n",
       "5631500400   180000.0       NaN\n",
       "6021501535   430000.0  700000.0\n",
       "7129300520   221900.0       NaN"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sales.pivot(index='property_id', columns='sale_number', values='price')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: perhaps demo pivot_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's have a look at our two remaining LTEE datasets. \n",
    "\n",
    "These two files record measurements of:\n",
    "\n",
    "- the cell size for every population for the first 10,000 generations, measured every 1000 generations\n",
    "- the relative fitness for every population for the first 10,000 generations, as measured by the growth rate of the strain relative to a reference strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These are TSV files, so specify tab as the delimiter\n",
    "cellsize = pd.read_csv('ltee_cell_size.tsv', sep=\"\\t\")\n",
    "fitness = pd.read_csv('ltee_relative_fitness.tsv', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gen</th>\n",
       "      <th>A-1</th>\n",
       "      <th>A-2</th>\n",
       "      <th>A-3</th>\n",
       "      <th>A-4</th>\n",
       "      <th>A-5</th>\n",
       "      <th>A-6</th>\n",
       "      <th>A+1</th>\n",
       "      <th>A+2</th>\n",
       "      <th>A+3</th>\n",
       "      <th>A+4</th>\n",
       "      <th>A+5</th>\n",
       "      <th>A+6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>500</td>\n",
       "      <td>0.4850</td>\n",
       "      <td>0.5100</td>\n",
       "      <td>0.4500</td>\n",
       "      <td>0.5150</td>\n",
       "      <td>0.5300</td>\n",
       "      <td>0.4550</td>\n",
       "      <td>0.4800</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>0.4850</td>\n",
       "      <td>0.5250</td>\n",
       "      <td>0.4600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000</td>\n",
       "      <td>0.5150</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>0.5800</td>\n",
       "      <td>0.6300</td>\n",
       "      <td>0.5150</td>\n",
       "      <td>0.4950</td>\n",
       "      <td>0.5550</td>\n",
       "      <td>0.5750</td>\n",
       "      <td>0.4950</td>\n",
       "      <td>0.5850</td>\n",
       "      <td>0.5850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2000</td>\n",
       "      <td>0.6000</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>0.6300</td>\n",
       "      <td>0.7000</td>\n",
       "      <td>0.7050</td>\n",
       "      <td>0.5450</td>\n",
       "      <td>0.5800</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6450</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5000</td>\n",
       "      <td>0.6650</td>\n",
       "      <td>0.9450</td>\n",
       "      <td>0.7000</td>\n",
       "      <td>0.7650</td>\n",
       "      <td>0.9150</td>\n",
       "      <td>0.6550</td>\n",
       "      <td>0.5950</td>\n",
       "      <td>0.8900</td>\n",
       "      <td>0.8500</td>\n",
       "      <td>0.6500</td>\n",
       "      <td>0.7250</td>\n",
       "      <td>0.7350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8000</td>\n",
       "      <td>0.6700</td>\n",
       "      <td>0.9400</td>\n",
       "      <td>0.6950</td>\n",
       "      <td>0.8150</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6950</td>\n",
       "      <td>0.8950</td>\n",
       "      <td>0.9750</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>0.8350</td>\n",
       "      <td>1.0800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9000</td>\n",
       "      <td>0.6850</td>\n",
       "      <td>0.9750</td>\n",
       "      <td>0.6850</td>\n",
       "      <td>0.8100</td>\n",
       "      <td>0.8800</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>0.7350</td>\n",
       "      <td>0.9200</td>\n",
       "      <td>0.9850</td>\n",
       "      <td>0.7900</td>\n",
       "      <td>0.8900</td>\n",
       "      <td>1.0200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9500</td>\n",
       "      <td>0.6700</td>\n",
       "      <td>0.9350</td>\n",
       "      <td>0.6850</td>\n",
       "      <td>0.7350</td>\n",
       "      <td>0.8500</td>\n",
       "      <td>0.6850</td>\n",
       "      <td>0.7200</td>\n",
       "      <td>0.9400</td>\n",
       "      <td>0.9650</td>\n",
       "      <td>0.8350</td>\n",
       "      <td>0.8950</td>\n",
       "      <td>0.9650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10000</td>\n",
       "      <td>0.6700</td>\n",
       "      <td>0.8450</td>\n",
       "      <td>0.7000</td>\n",
       "      <td>0.7600</td>\n",
       "      <td>0.7750</td>\n",
       "      <td>0.6550</td>\n",
       "      <td>0.7050</td>\n",
       "      <td>1.1000</td>\n",
       "      <td>0.8950</td>\n",
       "      <td>0.8450</td>\n",
       "      <td>0.8100</td>\n",
       "      <td>1.0150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gen     A-1     A-2     A-3     A-4     A-5     A-6     A+1     A+2  \\\n",
       "0      0  0.3725  0.3725  0.3725  0.3725  0.3725  0.3725  0.3725  0.3725   \n",
       "1    500  0.4850  0.5100  0.4500  0.5150  0.5300  0.4550  0.4800  0.4700   \n",
       "2   1000  0.5150  0.6250  0.6250  0.5800  0.6300  0.5150  0.4950  0.5550   \n",
       "3   2000  0.6000  0.6600  0.6300  0.7000  0.7050  0.5450  0.5800  0.6750   \n",
       "4   5000  0.6650  0.9450  0.7000  0.7650  0.9150  0.6550  0.5950  0.8900   \n",
       "5   8000  0.6700  0.9400  0.6950  0.8150  0.8700  0.6750  0.6950  0.8950   \n",
       "6   9000  0.6850  0.9750  0.6850  0.8100  0.8800  0.6600  0.7350  0.9200   \n",
       "7   9500  0.6700  0.9350  0.6850  0.7350  0.8500  0.6850  0.7200  0.9400   \n",
       "8  10000  0.6700  0.8450  0.7000  0.7600  0.7750  0.6550  0.7050  1.1000   \n",
       "\n",
       "      A+3     A+4     A+5     A+6  \n",
       "0  0.3725  0.3725  0.3725  0.3725  \n",
       "1  0.4700  0.4850  0.5250  0.4600  \n",
       "2  0.5750  0.4950  0.5850  0.5850  \n",
       "3  0.6450  0.6750  0.6750  0.6400  \n",
       "4  0.8500  0.6500  0.7250  0.7350  \n",
       "5  0.9750  0.7500  0.8350  1.0800  \n",
       "6  0.9850  0.7900  0.8900  1.0200  \n",
       "7  0.9650  0.8350  0.8950  0.9650  \n",
       "8  0.8950  0.8450  0.8100  1.0150  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellsize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To make life easier down the track, we'll rename the columns to match the variable and population names used in the `sampleruns` and `mutations` tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Generation</th>\n",
       "      <th>Ara-1</th>\n",
       "      <th>Ara-2</th>\n",
       "      <th>Ara-3</th>\n",
       "      <th>Ara-4</th>\n",
       "      <th>Ara-5</th>\n",
       "      <th>Ara-6</th>\n",
       "      <th>Ara+1</th>\n",
       "      <th>Ara+2</th>\n",
       "      <th>Ara+3</th>\n",
       "      <th>Ara+4</th>\n",
       "      <th>Ara+5</th>\n",
       "      <th>Ara+6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.3725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>500</td>\n",
       "      <td>0.4850</td>\n",
       "      <td>0.5100</td>\n",
       "      <td>0.4500</td>\n",
       "      <td>0.5150</td>\n",
       "      <td>0.5300</td>\n",
       "      <td>0.4550</td>\n",
       "      <td>0.4800</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>0.4850</td>\n",
       "      <td>0.5250</td>\n",
       "      <td>0.4600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000</td>\n",
       "      <td>0.5150</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>0.5800</td>\n",
       "      <td>0.6300</td>\n",
       "      <td>0.5150</td>\n",
       "      <td>0.4950</td>\n",
       "      <td>0.5550</td>\n",
       "      <td>0.5750</td>\n",
       "      <td>0.4950</td>\n",
       "      <td>0.5850</td>\n",
       "      <td>0.5850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2000</td>\n",
       "      <td>0.6000</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>0.6300</td>\n",
       "      <td>0.7000</td>\n",
       "      <td>0.7050</td>\n",
       "      <td>0.5450</td>\n",
       "      <td>0.5800</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6450</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>0.6400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5000</td>\n",
       "      <td>0.6650</td>\n",
       "      <td>0.9450</td>\n",
       "      <td>0.7000</td>\n",
       "      <td>0.7650</td>\n",
       "      <td>0.9150</td>\n",
       "      <td>0.6550</td>\n",
       "      <td>0.5950</td>\n",
       "      <td>0.8900</td>\n",
       "      <td>0.8500</td>\n",
       "      <td>0.6500</td>\n",
       "      <td>0.7250</td>\n",
       "      <td>0.7350</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Generation   Ara-1   Ara-2   Ara-3   Ara-4   Ara-5   Ara-6   Ara+1   Ara+2  \\\n",
       "0           0  0.3725  0.3725  0.3725  0.3725  0.3725  0.3725  0.3725  0.3725   \n",
       "1         500  0.4850  0.5100  0.4500  0.5150  0.5300  0.4550  0.4800  0.4700   \n",
       "2        1000  0.5150  0.6250  0.6250  0.5800  0.6300  0.5150  0.4950  0.5550   \n",
       "3        2000  0.6000  0.6600  0.6300  0.7000  0.7050  0.5450  0.5800  0.6750   \n",
       "4        5000  0.6650  0.9450  0.7000  0.7650  0.9150  0.6550  0.5950  0.8900   \n",
       "\n",
       "    Ara+3   Ara+4   Ara+5   Ara+6  \n",
       "0  0.3725  0.3725  0.3725  0.3725  \n",
       "1  0.4700  0.4850  0.5250  0.4600  \n",
       "2  0.5750  0.4950  0.5850  0.5850  \n",
       "3  0.6450  0.6750  0.6750  0.6400  \n",
       "4  0.8500  0.6500  0.7250  0.7350  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellsize.columns = (['Generation'] + ['Ara-{}'.format(n) for n  in range(1,7)] + \n",
    "                    ['Ara+{}'.format(n) for n  in range(1,7)])\n",
    "cellsize.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "fitness.columns = (['Generation'] + ['Ara-{}'.format(n) for n  in range(1,7)] + \n",
    "                    ['Ara+{}'.format(n) for n  in range(1,7)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise\n",
    "\n",
    "The cell size and relative fitness data is in wide form. Convert each one to tidy form. When thinking about which columns are \"wide\", it may help to aim to match the variables in the `sampleruns` and `mutations` tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging data "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can merge two datasets together by matching corresponding variables.\n",
    "\n",
    "Our main options are `DataFrame.join()` and `pandas.merge()`. `merge()` is a little more flexible, so we'll demonstrate that.\n",
    "\n",
    "Recall the `cars` dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpg</th>\n",
       "      <th>cylinders</th>\n",
       "      <th>displacement</th>\n",
       "      <th>horsepower</th>\n",
       "      <th>weight</th>\n",
       "      <th>acceleration</th>\n",
       "      <th>year</th>\n",
       "      <th>origin</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18.0</td>\n",
       "      <td>8</td>\n",
       "      <td>307.0</td>\n",
       "      <td>130</td>\n",
       "      <td>3504</td>\n",
       "      <td>12.0</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>chevrolet chevelle malibu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15.0</td>\n",
       "      <td>8</td>\n",
       "      <td>350.0</td>\n",
       "      <td>165</td>\n",
       "      <td>3693</td>\n",
       "      <td>11.5</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>buick skylark 320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18.0</td>\n",
       "      <td>8</td>\n",
       "      <td>318.0</td>\n",
       "      <td>150</td>\n",
       "      <td>3436</td>\n",
       "      <td>11.0</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>plymouth satellite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16.0</td>\n",
       "      <td>8</td>\n",
       "      <td>304.0</td>\n",
       "      <td>150</td>\n",
       "      <td>3433</td>\n",
       "      <td>12.0</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>amc rebel sst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17.0</td>\n",
       "      <td>8</td>\n",
       "      <td>302.0</td>\n",
       "      <td>140</td>\n",
       "      <td>3449</td>\n",
       "      <td>10.5</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>ford torino</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    mpg  cylinders  displacement  horsepower  weight  acceleration  year  \\\n",
       "0  18.0          8         307.0         130    3504          12.0    70   \n",
       "1  15.0          8         350.0         165    3693          11.5    70   \n",
       "2  18.0          8         318.0         150    3436          11.0    70   \n",
       "3  16.0          8         304.0         150    3433          12.0    70   \n",
       "4  17.0          8         302.0         140    3449          10.5    70   \n",
       "\n",
       "   origin                       name  \n",
       "0       1  chevrolet chevelle malibu  \n",
       "1       1          buick skylark 320  \n",
       "2       1         plymouth satellite  \n",
       "3       1              amc rebel sst  \n",
       "4       1                ford torino  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cars.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `origin` column here, which is a number 1-3, is actually intended to represent the country of origin. It's encoded as:\n",
    "\n",
    "- USA : 1\n",
    "- Europe : 2\n",
    "- Japan : 3\n",
    "\n",
    "Let's make a DataFrame to represent this mapping. We'll add a fourth code for Australia, which doesn't appear in the dataset!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>origin_country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Japan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   origin origin_country\n",
       "0       1            USA\n",
       "1       2         Europe\n",
       "2       3          Japan"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "origin_codes = pd.DataFrame(\n",
    "    {\n",
    "        'origin': [1,2,3],\n",
    "        'origin_country': ['USA','Europe','Japan']\n",
    "    }\n",
    ")\n",
    "origin_codes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use `pandas.merge()` to join our `cars` table to our `origin_codes` table using the shared `origin` field, like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpg</th>\n",
       "      <th>cylinders</th>\n",
       "      <th>displacement</th>\n",
       "      <th>horsepower</th>\n",
       "      <th>weight</th>\n",
       "      <th>acceleration</th>\n",
       "      <th>year</th>\n",
       "      <th>origin</th>\n",
       "      <th>name</th>\n",
       "      <th>origin_country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>22.0</td>\n",
       "      <td>6</td>\n",
       "      <td>198.0</td>\n",
       "      <td>95</td>\n",
       "      <td>2833</td>\n",
       "      <td>15.5</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>plymouth duster</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>46.6</td>\n",
       "      <td>4</td>\n",
       "      <td>86.0</td>\n",
       "      <td>65</td>\n",
       "      <td>2110</td>\n",
       "      <td>17.9</td>\n",
       "      <td>80</td>\n",
       "      <td>3</td>\n",
       "      <td>mazda glc</td>\n",
       "      <td>Japan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>29.8</td>\n",
       "      <td>4</td>\n",
       "      <td>89.0</td>\n",
       "      <td>62</td>\n",
       "      <td>1845</td>\n",
       "      <td>15.3</td>\n",
       "      <td>80</td>\n",
       "      <td>2</td>\n",
       "      <td>vokswagen rabbit</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>33.8</td>\n",
       "      <td>4</td>\n",
       "      <td>97.0</td>\n",
       "      <td>67</td>\n",
       "      <td>2145</td>\n",
       "      <td>18.0</td>\n",
       "      <td>80</td>\n",
       "      <td>3</td>\n",
       "      <td>subaru dl</td>\n",
       "      <td>Japan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>15.5</td>\n",
       "      <td>8</td>\n",
       "      <td>318.0</td>\n",
       "      <td>145</td>\n",
       "      <td>4140</td>\n",
       "      <td>13.7</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>dodge monaco brougham</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>14.0</td>\n",
       "      <td>8</td>\n",
       "      <td>318.0</td>\n",
       "      <td>150</td>\n",
       "      <td>4237</td>\n",
       "      <td>14.5</td>\n",
       "      <td>73</td>\n",
       "      <td>1</td>\n",
       "      <td>plymouth fury gran sedan</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>15.0</td>\n",
       "      <td>8</td>\n",
       "      <td>390.0</td>\n",
       "      <td>190</td>\n",
       "      <td>3850</td>\n",
       "      <td>8.5</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>amc ambassador dpl</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>23.0</td>\n",
       "      <td>4</td>\n",
       "      <td>140.0</td>\n",
       "      <td>83</td>\n",
       "      <td>2639</td>\n",
       "      <td>17.0</td>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>ford pinto</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>14.0</td>\n",
       "      <td>8</td>\n",
       "      <td>454.0</td>\n",
       "      <td>220</td>\n",
       "      <td>4354</td>\n",
       "      <td>9.0</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>chevrolet impala</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>15.0</td>\n",
       "      <td>8</td>\n",
       "      <td>304.0</td>\n",
       "      <td>150</td>\n",
       "      <td>3892</td>\n",
       "      <td>12.5</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>amc matador (sw)</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      mpg  cylinders  displacement  horsepower  weight  acceleration  year  \\\n",
       "14   22.0          6         198.0          95    2833          15.5    70   \n",
       "296  46.6          4          86.0          65    2110          17.9    80   \n",
       "385  29.8          4          89.0          62    1845          15.3    80   \n",
       "299  33.8          4          97.0          67    2145          18.0    80   \n",
       "148  15.5          8         318.0         145    4140          13.7    77   \n",
       "67   14.0          8         318.0         150    4237          14.5    73   \n",
       "9    15.0          8         390.0         190    3850           8.5    70   \n",
       "117  23.0          4         140.0          83    2639          17.0    75   \n",
       "6    14.0          8         454.0         220    4354           9.0    70   \n",
       "53   15.0          8         304.0         150    3892          12.5    72   \n",
       "\n",
       "     origin                      name origin_country  \n",
       "14        1           plymouth duster            USA  \n",
       "296       3                 mazda glc          Japan  \n",
       "385       2          vokswagen rabbit         Europe  \n",
       "299       3                 subaru dl          Japan  \n",
       "148       1     dodge monaco brougham            USA  \n",
       "67        1  plymouth fury gran sedan            USA  \n",
       "9         1        amc ambassador dpl            USA  \n",
       "117       1                ford pinto            USA  \n",
       "6         1          chevrolet impala            USA  \n",
       "53        1          amc matador (sw)            USA  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cars_withorigin = pd.merge(cars, origin_codes)\n",
    "cars_withorigin.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This \"just worked\" because Pandas correctly deduced that the identically-named field(s) were the ones to match on. Sometimes we might need to be more verbose. In this case, this accomplishes the same thing as the above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use origin from the left dataframe (cars) and from the right (origin_codes)\n",
    "# use how=\"left\" (keep all origin values that exist in the left dataframe)\n",
    "cars_withorigin = pd.merge(cars, origin_codes, left_on='origin', right_on='origin', how=\"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try to merge our `sampleinfo` and `mutations` columns. This time there are three shared fields: 'Strain ID', 'Population', and 'Generation'. In fact only 'Strain ID' is needed to uniquely identify  rows, but we want to specify all matching variables so that Pandas knows to only include each of these variables once in the resulting dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Population', 'Generation', 'Strain ID', 'Mutator', 'Reference',\n",
       "       'Facility', 'Accession', 'Read Type', 'Read Length', 'Sequencing Depth',\n",
       "       'Analysis Notes', 'Total Mutations', 'Base Substitutions',\n",
       "       'IS Element Insertions', 'Small Indels', 'Large Deletions',\n",
       "       'Large Duplications', 'Total Deleted Base Pairs',\n",
       "       'Total Inserted Base Pairs', 'Estimated Final Genome Size',\n",
       "       'Synonymous Base Substitutions', 'Nonsynonymous Base Substitutions',\n",
       "       'Nonsense Base Substitutions', 'Pseudogene Base Substitutions',\n",
       "       'Noncoding RNA Base Substitutions', 'Intergenic Base Substitutions'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ltee = pd.merge(sampleruns, mutations, on=['Strain ID','Population','Generation'])\n",
    "ltee.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Strain ID</th>\n",
       "      <th>Mutator</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Facility</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Read Type</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Sequencing Depth</th>\n",
       "      <th>...</th>\n",
       "      <th>Large Duplications</th>\n",
       "      <th>Total Deleted Base Pairs</th>\n",
       "      <th>Total Inserted Base Pairs</th>\n",
       "      <th>Estimated Final Genome Size</th>\n",
       "      <th>Synonymous Base Substitutions</th>\n",
       "      <th>Nonsynonymous Base Substitutions</th>\n",
       "      <th>Nonsense Base Substitutions</th>\n",
       "      <th>Pseudogene Base Substitutions</th>\n",
       "      <th>Noncoding RNA Base Substitutions</th>\n",
       "      <th>Intergenic Base Substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584408</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>70.241277</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>6244</td>\n",
       "      <td>0</td>\n",
       "      <td>4623568</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>500</td>\n",
       "      <td>REL768B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584409</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>41.136778</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4772</td>\n",
       "      <td>1349</td>\n",
       "      <td>4626389</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584410</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>33.226052</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2544</td>\n",
       "      <td>0</td>\n",
       "      <td>4627268</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1000</td>\n",
       "      <td>REL958B</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584411</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>34.589932</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1446</td>\n",
       "      <td>4631258</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>1500</td>\n",
       "      <td>REL1062A</td>\n",
       "      <td>None</td>\n",
       "      <td>This study</td>\n",
       "      <td>Genoscope</td>\n",
       "      <td>SRR2584438</td>\n",
       "      <td>single</td>\n",
       "      <td>36</td>\n",
       "      <td>40.342099</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>6305</td>\n",
       "      <td>0</td>\n",
       "      <td>4623507</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Population  Generation Strain ID Mutator   Reference   Facility   Accession  \\\n",
       "0      Ara+1         500   REL768A    None  This study  Genoscope  SRR2584408   \n",
       "1      Ara+1         500   REL768B    None  This study  Genoscope  SRR2584409   \n",
       "2      Ara+1        1000   REL958A    None  This study  Genoscope  SRR2584410   \n",
       "3      Ara+1        1000   REL958B    None  This study  Genoscope  SRR2584411   \n",
       "4      Ara+1        1500  REL1062A    None  This study  Genoscope  SRR2584438   \n",
       "\n",
       "  Read Type  Read Length  Sequencing Depth              ...                \\\n",
       "0    single           36         70.241277              ...                 \n",
       "1    single           36         41.136778              ...                 \n",
       "2    single           36         33.226052              ...                 \n",
       "3    single           36         34.589932              ...                 \n",
       "4    single           36         40.342099              ...                 \n",
       "\n",
       "  Large Duplications  Total Deleted Base Pairs  Total Inserted Base Pairs  \\\n",
       "0                  0                      6244                          0   \n",
       "1                  0                      4772                       1349   \n",
       "2                  0                      2544                          0   \n",
       "3                  0                         0                       1446   \n",
       "4                  0                      6305                          0   \n",
       "\n",
       "   Estimated Final Genome Size  Synonymous Base Substitutions  \\\n",
       "0                      4623568                              0   \n",
       "1                      4626389                              0   \n",
       "2                      4627268                              0   \n",
       "3                      4631258                              0   \n",
       "4                      4623507                              0   \n",
       "\n",
       "   Nonsynonymous Base Substitutions  Nonsense Base Substitutions  \\\n",
       "0                                 2                            0   \n",
       "1                                 1                            0   \n",
       "2                                 2                            0   \n",
       "3                                 1                            0   \n",
       "4                                 4                            0   \n",
       "\n",
       "   Pseudogene Base Substitutions  Noncoding RNA Base Substitutions  \\\n",
       "0                              0                                 0   \n",
       "1                              0                                 0   \n",
       "2                              0                                 0   \n",
       "3                              0                                 0   \n",
       "4                              0                                 0   \n",
       "\n",
       "   Intergenic Base Substitutions  \n",
       "0                              1  \n",
       "1                              0  \n",
       "2                              0  \n",
       "3                              0  \n",
       "4                              1  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ltee.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the metadata on Mutator phenotypes is together with the information on actual mutations, we could try exploring the relationships between these fields. Here are a couple of previews of ways to do this!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdbf2346e80>"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEKCAYAAAAFJbKyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAGOBJREFUeJzt3X+w3XV95/HnKxAqiitgIosJEcdE+8OlFLLojtqh4xAJVtG2TnFczbrOgCsaXbWKnXFl7fijP7DbWOuIC+3FsTB0bbexQwop1bW21ST8aACp5opUL1AggiIFMcl97x/ne+EEbs4933DPPefkPh8zZ845n/P5fs/7cof7yuf74/NJVSFJUr+WDLsASdJ4MTgkSa0YHJKkVgwOSVIrBockqRWDQ5LUisEhSWrF4JAktWJwSJJaOXzYBQzCsmXL6sQTTxx2GZI0Vq677rrdVbV8rn6HZHCceOKJ7NixY9hlSNJYSfIv/fTzUJUkqRWDQ5LUisEhSWrF4JAktWJwSFIfdu/ezTve8Q6+//3vD7uUoTM4JKkPExMT7Ny5k4mJiWGXMnQGhyTNYffu3WzZsoWqYsuWLYt+1GFwSNIcJiYmmFlme3p6etGPOgwOSZrD1q1b2bNnDwB79uzhmmuuGXJFwzWw4EhyQpIvJbk1yS1J3tm0X5jkjiQ3No+zurb5QJLJJN9M8oqu9jObtskkFwyqZkmazRlnnMHSpUsBWLp0KevWrRtyRcM1yBHHXuA9VfUzwIuB85P8bPPZ71fVyc3jKoDms3OAnwPOBP4oyWFJDgM+BawHfhZ4fdd+JGngNmzYQBIAlixZwoYNG4Zc0XANLDiq6q6qur55/SPgVmBFj03OBq6oqkeq6jvAJHBa85isqtuq6ifAFU1fSVoQy5YtY/369SRh/fr1PPOZzxx2SUO1IOc4kpwI/ALw9abp7Ul2Jrk0yTFN2wrge12bTTVtB2qXpAWzYcMGTjrppEU/2oAFCI4kRwFfAN5VVQ8AnwaeB5wM3AVcNNN1ls2rR/vjv+fcJDuS7Lj33nvnpXZJmrFs2TI++clPLvrRBgw4OJIspRMan6+qPweoqrural9VTQOfpXMoCjojiRO6Nl8J3NmjfT9VdXFVra2qtcuXzzmdvCTpIA3yqqoAlwC3VtUnutqP7+r2WuDm5vVm4JwkP5XkucAaYBuwHViT5LlJjqBzAn3zoOqWJPU2yIWcXgK8EbgpyY1N22/SuSrqZDqHm24HzgOoqluSXAl8g84VWedX1T6AJG8HrgYOAy6tqlsGWLckqYfM3A15KFm7dm25AqAktZPkuqpaO1c/7xyXJLVicEiSWjE4JEmtGBySpFYMDklSKwaHJKkVg0OS1IrBIUl92L17N+94xzsW/bKxYHBIUl8mJibYuXPnol82FgwOSZrT7t272bJlC1XFli1bFv2ow+CQpDlMTEwwMz3T9PT0oh91GBySNIetW7eyZ88eAPbs2cM111wz5IqGy+CQpDmcccYZLF26FIClS5eybt26IVc0XAaHJM1hw4YNdJYYgiVLliz65WMNDkmaw7Jly1i/fj1JWL9+/aJfPtbgkKQ+vOpVr+KpT30qr371q4ddytAZHJLUhy9+8Ys89NBDbN7sytUGhyTNwfs49mdwSNIcvI9jfwaHJM3B+zj2Z3BI0hy8j2N/BockzcH7OPZncEjSHLyPY3+HD7sASRoHGzZs4Pbbb1/0ow0wOCSpL8uWLeOTn/zksMsYCR6qkiS1YnBIkloxOCRJrRgckqRWDA5JUisDC44kJyT5UpJbk9yS5J1N+7FJtibZ1Twf07QnyaYkk0l2Jjmla18bmv67kngtnCQN0SAvx90LvKeqrk/ydOC6JFuB/wJcW1UfT3IBcAHwfmA9sKZ5vAj4NPCiJMcCHwLWAtXsZ3NV3T/A2iWNqU2bNjE5OTnv+52amgJg5cqV875vgNWrV7Nx48aB7Hu+DWzEUVV3VdX1zesfAbcCK4CzgZmpJSeA1zSvzwYuq46vAUcnOR54BbC1qu5rwmIrcOag6pak2Tz88MM8/PDDwy5jJCzIDYBJTgR+Afg6cFxV3QWdcEnyrKbbCuB7XZtNNW0Han/8d5wLnAuwatWq+f0BJI2NQf2rfWa/mzZtGsj+x8nAT44nOQr4AvCuqnqgV9dZ2qpH+/4NVRdX1dqqWrt8+fKDK1aSNKeBBkeSpXRC4/NV9edN893NISia53ua9inghK7NVwJ39miXJA3BIK+qCnAJcGtVfaLro83AzJVRG4C/7Gp/U3N11YuBHzaHtK4G1iU5prkCa13TJkkagkGe43gJ8EbgpiQ3Nm2/CXwcuDLJW4DvAq9rPrsKOAuYBB4C3gxQVfcl+S1ge9Pvw1V13wDrliT1MLDgqKqvMvv5CYCXz9K/gPMPsK9LgUvnrzpJ0sHyznFJUisGhySpFYNDktSKwSFJasXgkCS1YnBIkloxOCRJrRgckqRWDA5JUisGhySpFYNDktSKwSFJasXgkCS1MmdwJHl7kn/XvP5Mkm1JnjC7rSRpcehnxHFuVT2QZB2dtb7/G/A7gy1LkjSq+gmOmfW91wN/XFXX9bmdJOkQ1E8A/FOSq4BXAVuSHMVjYSJJWmT6WQHwzcCpwGRVPZRkGfCWwZYlSRpVcwZHVe1L8l1gdZJBrlEuSRoDcwZBko8C/xn4Z2Bf01zAWQOsS5I0ovoZQfwq8Pyq+vGgi5Ekjb5+To5/p89+kqRFoJ8Rx4+AG5L8DfDITGNVvXtgVUmSRlY/wfHXzUOSpL6uqrqkuZpqddM0WVV7B1uWJGlU9XNV1cuAzwF3AAH+fZI3VtXfD7o4SdLo6edQ1e8DZ1XVNwCS/AydIFk7yMIkSaOpn6uljpgJDYCquhU4YnAlSZJGWT8jjuuTfIbOKAPgDcANgytJkjTK+hlxvBX4NvA+4P3AbcB5c22U5NIk9yS5uavtwiR3JLmxeZzV9dkHkkwm+WaSV3S1n9m0TSa5oM0PJ0maf/1cVfVjOutvtF2D40+APwQue1z771fV73U3JPlZ4Bzg54BnA3+T5PnNx58CzgCmgO1JNncfOpMkLawDBkeSy6vq9UluYJZp1KvqlF47rqqvJDmxzzrOBq6oqkeA7ySZBE5rPpusqtuamq5o+hockjQkvUYcv9E8/9o8f+fbk7wJ2AG8p6rup7Oy4Ne6+kw1bQDfe1z7i+a5HklSCwc8x1FVU83Lt1TVt7sfHPx6HJ8GngecDNwFXNS0Z7YSerQ/QZJzk+xIsuPee+89yPIkSXPp5+T4mbO0vfJgvqyq7q6qfVU1DXyWxw5HTQEndHVdCdzZo322fV9cVWurau3y5csPpjxJUh8OGBxJzmvOb7wgyfVdj13ArQfzZUmO73r7WmDmiqvNwDlJfirJc4E1wDZgO7AmyXOTHEHnBPrmg/luSdL86HWO40rgWuBjQPdlsD+qqnvm2nGSy4HTgWVJpoAPAacnOZnO4abbaS7rrapbklxJ56T3XuD8qtrX7OftwNXAYcClVXVLmx9QkjS/DhgczUnr+4HXASQ5FngKcHiSZ1fVrIeMurZ//SzNl/To/xHgI7O0XwVc1eu7JEkLZ85zHEnOSvItOucbvk7nKqe/HXRhkqTR1M/J8Y8CLwG+WVUn0DlZ/uVBFiVJGl39BMfeqroXWJIkVbUV6HnznyTp0NXPJIc/TPI04KvAZUnuAaYHW5YkaVT1M+J4DfBj4F10DlHdAfzyAGuSJI2wfoLjA81Ne3uq6pKq+gTw7kEXJkkaTQt657gkafz1mh33PDprcTw/yfVdHz0duG7QhUmSRtPA7hyXJB2a5rxzPMk7H79NP3eOS5IOTf1cjnstj01x/hQ6s9V+G3jBAOuSJI2ofpaO/Znu90lOA948sIokSSOtn6uq9lNV23hsHQ1J0iIz54gjycaut0uAU4H7BlaRJGmk9XOOo3s5vb3A3wB/NphyJEmjrp9zHB9ciEIkSeOh1w2Af95rw6r6lfkvR5I06nqNOE6ns7zr5XTuFM8C1CNJGnG9guM44BXA65vHZuDyqvrmQhQmSRpNB7wct5kN96+q6g10VgD8LvDVJG9bsOokSSOn58nxJEuB9XRGHM8H/gj44gLUJUkaUb1Ojl9CZ4nYq4HfrqobF6wqSdLI6jXieDPwAHAecG7y6LnxAFVVxw64NknSCOoVHEsXrApJi86mTZuYnJwcdhl927VrFwAbN26co+doWb169bzX3Gta9X3z+k2S1GVycpIbbrkBjh52JX2a7jzdcMcNw62jjR8MZrf9TDkiSYNxNEyfPj3sKg5ZS77ceh7b/vY7kL1Kkg5ZBockqZVel+PeT2flvyd8hFdVSdKi1escx7IFq0KSNDZ6TTmyr/sBPIPO/FUzj56SXJrkniQ3d7Udm2Rrkl3N8zFNe5JsSjKZZGeSU7q22dD035Vkw5P5YSVJT96c5ziSvDLJt4Ap4OvN89/2se8/Ac58XNsFwLVVtQa4tnkPnWlN1jSPc4FPN999LPAh4EV0lqv90EzYSJKGo5+T4x+hM8nhN6vqBDoz5n55ro2q6is8cYnZs4GJ5vUE8Jqu9suq42vA0UmOb75ra1XdV1X3A1t5YhhJkhZQP8Gxt6ruBZYkSVVtpTOH1cE4rqruAmien9W0rwC+19Vvqmk7ULskaUj6uQHwh0meBnwVuCzJPTx6D+W8mW2RqOrR/sQdJOfSOczFqlWr5q8ySdJ++hlxvAb4MfAuOoeo7gB++SC/7+7mEBTN8z1N+xRwQle/lcCdPdqfoKourqq1VbV2+fLlB1meJGku/QTHB5orq/ZU1SVV9Qng3Qf5fZuBmSujNgB/2dX+pubqqhcDP2wOZV0NrEtyTHNSfF3TJkkakn6CY7aT0a+ca6MklwP/CLwgyVSStwAfB85Isgs4o3kPcBVwGzAJfBZ4G0BV3Qf8FrC9eXy4aZMkDUmvO8fPA94KPD/J9V0fPR3YMdeOq+r1B/jo5bP0LeD8A+znUuDSub5PkrQwep0cv5LOvRYf47H7LQB+VFX3zL6JJOlQ12s9jvuB+4HXJXkh8NLmo7/jsZPakqRFpp87x8+nM/pY1TyuTPK2QRcmSRpN/dzHcR5wWlU9CJDko8A/AH80yMIkSaOpn6uqAuzper+H2W/MkyQtAr2uqjq8qvYCnwO+luQLzUev5bH5piTpoExNTcEPB7e8qYAfwFRNzftuex2q2gacUlW/k+RLwMvojDTeWlXb570SSdJY6BUcjx6OaoLCsJA0b1auXMm9uZfp0+d76jvNWPLlJaxcsXLe99srOJYnOeDUIs3UI5KkRaZXcBwGHIUnwiVJXXoFx11V9eEFq0SSNBZ6Xc7gSEOS9AS9guMJkxFKknTA4HD6cknSbLzzRpLUisEhSWrF4JAktWJwSJJaMTgkSa0YHJKkVgwOSVIrBockqRWDQ5LUisEhSWrF4JAktWJwSJJaMTgkSa0YHJKkVgwOSVIrBockqZWhBEeS25PclOTGJDuatmOTbE2yq3k+pmlPkk1JJpPsTHLKMGqWJHUMc8TxS1V1clWtbd5fAFxbVWuAa5v3AOuBNc3jXODTC16pJOlRhw+7gC5nA6c3ryeALwPvb9ovq6oCvpbk6CTHV9VdQ6lS0vz5ASz58pgcMX+weT5qqFW08wNgxfzvdljBUcA1SQr4TFVdDBw3EwZVdVeSZzV9VwDf69p2qmnbLziSnEtnRMKqVasGXL6kJ2v16tXDLqGVXbt2AbBmxZohV9LCisH8dx5WcLykqu5swmFrkn/u0TeztNUTGjrhczHA2rVrn/C5pNGycePGYZfQyky9mzZtGnIlwzeUMWJV3dk83wP8BXAacHeS4wGa53ua7lPACV2brwTuXLhqJUndFjw4kjwtydNnXgPrgJuBzcCGptsG4C+b15uBNzVXV70Y+KHnNyRpeIZxqOo44C+SzHz/n1bVXyfZDlyZ5C3Ad4HXNf2vAs4CJoGHgDcvfMmSpBkLHhxVdRvw87O0fx94+SztBZy/AKVJkvowJtfBSZJGhcEhSWrF4JAktWJwSJJaMTgkSa0YHJKkVgwOSVIrBockqRWDQ5LUisEhSWrF4JAktWJwSJJaMTgkSa0YHJKkVgwOSVIrBockqRWDQ5LUisEhSWrF4JAktWJwSJJaMTgkSa0YHJKkVg4fdgFqZ9u2bbzvfe/joosu4tRTTx12OdLI2bRpE5OTk/O+3127dgGwcePGed83wOrVqwe27/lmcIyZCy+8kOnpaT74wQ9y1VVXDbucQ9ag/vhMTU0BsHLlynnf9zj94RlHRx555LBLGBkGxxjZtm0bDz74IAAPPvgg1113naOOMfPwww8Pu4RDnuE5eKmqYdcw79auXVs7duwYdhnz7qyzzno0OACOOuqoRT3qGNSoYJBmDnesWbNmyJX0z5HM4pHkuqpaO1c/RxwDMog/at2hMfN+vv+HHqc/EpOTk3zr5utZddS+YZfStyP2dK5H+fHt24dcSX++++Bhwy5BI8jgGCNLlixhenp6v/eL2dTUFOM2YD7uqdNzdxohVY+dl5FmGBwDMoh/tW/bto33vve9j773yipJwzA2wZHkTOAPgMOA/11VH5+P/Y7jcXLojDYmJiaYmJgYdil9GcQhsJUrV/KtH9wzr/sctLsf6owSx2XkkQzmCjCNt7EIjiSHAZ8CzgCmgO1JNlfVN57svicnJ7nhpm8w/dRjn+yuFkSW/BSZfoQ9RzyD6779r8Mupy9LHrpvIPtdvXr1QPY7SD9pTo4/5cTxODn+fMbzv7MGayyCAzgNmKyq2wCSXAGcDTzp4Ogcvx2fA+X1tGPGqNoZNZDj5IM8iT+OI9FxurBB421cgmMF8L2u91PAi+Zt7/v2suSh78/b7gCY3sfYnbmFzrGJJfN8Jc2+vfO7vzHmTWQ6FIxLcGSWtv3+Kic5FzgXYNWqVX3v+PTTTx/YHcLjeLPXkUceObC7mseJ/3KXDmxcgmMKOKHr/Urgzu4OVXUxcDF0bgDsd8f+gZCkdsblRoDtwJokz01yBHAOsHnINUnSojQWI46q2pvk7cDVdC7HvbSqbhlyWZK0KI1FcABU1VXA4p2YSZJGxLgcqpIkjQiDQ5LUisEhSWrF4JAktWJwSJJaOSRXAExyL/Avw65jgJYBu4ddhA6av7/xdaj/7p5TVcvn6nRIBsehLsmOfpZ31Gjy9ze+/N11eKhKktSKwSFJasXgGE8XD7sAPSn+/saXvzs8xyFJaskRhySpFYNjBCSpJBd1vX9vkguHWJLmUZIHm+clSTYluTnJTUm2J3nusOvTEyXZl+TG5nf1Z0me2rRXks919Ts8yb1J/mp41S48g2M0PAL8SpJlwy5EA/XrwLOBk6rqPwCvBX4w3JJ0AA9X1clV9ULgJ8Bbm/Z/A16YZGYN4DOAO4ZR4DAZHKNhL52Tbv/98R8keU6Sa5PsbJ5XNe1/0vzr9R+S3Jbk17q2+Y3mX7M7k/zPhfsxNIfjgbuqahqgqqaq6v4h16S5/R3QvfbxFuCVzevXA5cveEVDZnCMjk8Bb0jyjMe1/yFwWVWdBHwe2NT12fHAS4FfBj4OkGQdsAY4DTgZODXJLw64dvXnSuBVzSGQi5L8wrALUm9JDgfWAzd1NV8BnJPkKcBJwNeHUdswGRwjoqoeAC4DHr8I+n8C/rR5/Tk6QTHj/1bVdFV9AziuaVvXPG4Argd+mk6QaMiqagp4AfABYBq4NsnLh1uVDuDIJDcCO4DvApfMfFBVO4ET6Yw2FuXicmOzAuAi8b/o/LH/4x59uq+ffqTrdbqeP1ZVn5nn2jQPquoROoc6tiS5G3gNcO1wq9IsHq6qk3t8vhn4PeB04JkLUtEIccQxQqrqPjqHM97S1fwPwDnN6zcAX51jN1cD/zXJUQBJViR51nzXqvaSnJLk2c3rJXQOcxzKk3Eeyi4FPlxVN83Z8xBkcIyei+jMwDljI/DmJDuBNwLv7LVxVV1D59DWPya5Cfg/wNMHVKvaeRbwxSQ3AzvpXBTxh8MtSQejubDhD4Zdx7B457gkqRVHHJKkVgwOSVIrBockqRWDQ5LUisEhSWrF4JD6cLCzoiY5Osnb+th/X/2kUWBwSP052FlRjwb6CYR++z0qyWFt+kvzxeCQ+nfAWVGTXJjkvV3vb05yIp3JJ5/XTGz4u0mOamY5vr5Zk+PsZpPH90vzPLN2x683+z09yZeS/Cn7T7wnLRjnqpL6dwXwP5rDUyfRmXbiZXNscwHwwpl5j5rZVl9bVQ806698LcnmWfr9Kp3ZjX+ezkwC25N8pdnnaU3f78zvjyf1x+CQ+lRVO5tRxJOZFTXAR5up7qeBFTw2s3G3lwKXV9U+4O4k/w/4j8ADwDZDQ8NkcEjtHGhW1L3sf+j3KQfY/g3AcuDUqtqT5PYD9M0sbTP+rd9ipUHwHIfUzoFmRb0dOAU6s+ACM2uJ/4j9J5l8BnBPExq/BDznAP2+Avx6ksOSLAd+Edg2nz+IdLAccUgtNIsxzTYr6heANzWL/2wHvtX0/36Sv29mxN0C/DadGXJ3ADcC/3yAfu+js4jXP9FZg+V9VfWvSX56sD+hNDdnx5UkteKhKklSKwaHJKkVg0OS1IrBIUlqxeCQJLVicEiSWjE4JEmtGBySpFb+P+7OT/aA97/TAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.boxplot(data=ltee, x='Mutator', y='Total Mutations')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mutator\n",
       "IS       66.625000\n",
       "None     21.449198\n",
       "PM      921.262295\n",
       "Name: Total Mutations, dtype: float64"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ltee.groupby('Mutator')['Total Mutations'].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise\n",
    "\n",
    "Merge the cell size data on to the `ltee` table with `pd.merge()`, using the Generation and Population variables. This is only possible if it's been converted to tidy form first, with variable names and values corresponding to those in `1tee`! You will probably want to set `how=\"left\"` (if the cell size data is on the right), since the `ltee` table contains generations that don't exist in the cell size data, and we don't want to throw these rows away.\n",
    "\n",
    "Do the same with the relative fitness data.\n",
    "\n",
    "Have a look at missing and repeated values in the newly-added columns. What do you see?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing to a file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To finish off, let's write out our tidied and merged table to a new file, for future analyses. To write to CSV, we can use the `to_csv` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Don't keep the index as we still have the Strain ID column\n",
    "ltee.to_csv('ltee_solution.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "138px",
    "width": "252px"
   },
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
